pymc_marketing/clv/models/shifted_beta_geo.py

```diff
@@ -77,59 +76,81 @@ class ShiftedBetaGeoModelIndividual(CLVModel):
+        sampler_config: Optional[Dict] = None,
     ):
-        super().__init__()
-
-        t_churn = np.asarray(t_churn)
-        T = np.asarray(T)
-
-        if np.any((t_churn < 0) | (t_churn > T) | np.isnan(t_churn)):
+        try:
+            self.customer_id = data["customer_id"]
+        except KeyError:
+            raise KeyError("data must contain a 'customer_id' column")
+        try:
+            self.t_churn: np.ndarray = np.asarray(data["t_churn"])
+        except KeyError:
+            raise KeyError("data must contain a 't_churn' column")
+        try:
+            self.T: np.ndarray = np.asarray(data["T"])
+        except KeyError:
+            raise KeyError("data must contain a 'T' column")
+        super().__init__(model_config=model_config, sampler_config=sampler_config)
+        self.data = data
+        self.alpha_prior = self.create_distribution_from_prior(
+            self.model_config["alpha_prior"]["dist"],
+            **self.model_config["alpha_prior"]["kwargs"]
+        )
+        self.beta_prior = self.create_distribution_from_prior(
+            self.model_config["beta_prior"]["dist"],
+            **self.model_config["beta_prior"]["kwargs"]
+        )
+        self._process_priors(self.alpha_prior, self.beta_prior)
+
+        if np.any(
+            (self.t_churn < 0) | (self.t_churn > self.T) | np.isnan(self.t_churn)
+        ):
             raise ValueError(
                 "t_churn must respect 0 < t_churn <= T.\n",
                 "Customers that are still alive should have t_churn = T",
             )
+        self.coords = {"customer_id": np.asarray(self.customer_id)}
 
-        alpha_prior, beta_prior = self._process_priors(alpha_prior, beta_prior)
+    @property
+    def default_model_config(self) -> Dict:
+        return {
+            "alpha_prior": {"dist": "halfflat", "kwargs": {}},
+            "beta_prior": {"dist": "halfflat", "kwargs": {}},
+        }
 
-        coords = {"customer_id": np.asarray(customer_id)}
-        with pm.Model(coords=coords) as self.model:
-            alpha = self.model.register_rv(alpha_prior, name="alpha")
-            beta = self.model.register_rv(beta_prior, name="beta")
+    @property
+    def _serializable_model_config(self) -> Dict:
+        return self.model_config
+
+    @property
+    def default_sampler_config(self) -> Dict:
+        return {}
+
+    def build_model(
+        self,
+    ) -> None:
+        with pm.Model(coords=self.coords) as self.model:
+            alpha = self.model.register_rv(self.alpha_prior, name="alpha")
+            beta = self.model.register_rv(self.beta_prior, name="beta")
 
             theta = pm.Beta("theta", alpha, beta, dims=("customer_id",))
 
             churn_raw = pm.Geometric.dist(theta)
-
             pm.Censored(
                 "churn_censored",
                 churn_raw,
                 lower=None,
-                upper=T,
-                observed=t_churn,
+                upper=self.T[0],
```

Comment:
Why is this `T[0]`?

Reply:
as previously the T was passed as a standalone value it was used as an int. Now as I'm passing dataset of input params, it's value is propagated as a value of each row of parameters, with just self.T the pm.Censored upper limit was defined as <constant> in the str representation

Reply:
I think T is allowed to be a vector per observation, although I guess we didn't test it if it didn't show up as a failure.

Reply:
Yeah the docstrings say that explicitly. Need to fix. 

If passing a single value is no longer possible because everything must be a DataFrame, the tests and docstrings must be updated.

Reply:
I've adapted the docstrings regarding expected input parameters, please check again

Reply:
Sounds like you should pass the whole column? There is no constraint that all `T` must be the same. Before we allowed broadcasting if a user passed a scalar but they could also have passed a vector with different values.

Reply:
it doesn't make the model fail, but the str. representation will need to be adapted to churn_censored ~Censored(Geometric(theta), -inf,<constant>) as I assume that PyMC also assumes that in case of an array the number doesn't need to be the same

Reply:
I'll adapt the test and make it again entire T